pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,475 @@
+		#[pallet::weight(T::WeightInfo::claim_reward())]
+		#[transactional]
+		pub fn claim_reward(origin: OriginFor<T>, account_id: T::AccountId) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			T::Rewards::claim_reward((T::Domain::get(), STAKE_CURRENCY_ID), &account_id).map(|_| ())
+		}
+
+		/// Admin method to set the reward amount for a collator used for the next sessions.
+		/// Current session is not affected by this call.
+		#[pallet::weight(T::WeightInfo::set_collator_reward())]
+		pub fn set_collator_reward(
+			origin: OriginFor<T>,
+			collator_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextSessionChanges::<T>::mutate(|changes| {
+				changes.collator_reward = Some(collator_reward_per_session);
+			});
+
+			Ok(())
+		}
+
+		/// Admin method to set the total reward distribution for the next sessions.
+		/// Current session is not affected by this call.
+		///
+		/// Throws if total_reward < collator_reward * collator_count.
+		#[pallet::weight(T::WeightInfo::set_total_reward())]
+		pub fn set_total_reward(
+			origin: OriginFor<T>,
+			total_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextSessionChanges::<T>::try_mutate(|changes| {
+				let current = ActiveSessionData::<T>::get();
+				let total_collator_rewards = changes
+					.collator_reward
+					.unwrap_or(current.collator_reward)
+					.saturating_mul(
+						changes
+							.collator_count
+							.unwrap_or(current.collator_count)
+							.into(),
+					);
+				ensure!(
+					total_reward_per_session >= total_collator_rewards,
+					Error::<T>::InsufficientTotalReward
+				);
+
+				changes.total_reward = Some(total_reward_per_session);
+				Ok(())
+			})
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Mint default amount of stake for target address and deposit stake.
+	/// Enables receiving rewards onwards.
+	///
+	/// Weight (6 reads, 6 writes):
+	///  * mint_into (2 reads, 2 writes): Account, TotalIssuance
+	///  * deposit_stake (4 reads, 4 writes): Currencies, Groups, StakeAccounts, Account
+	pub(crate) fn do_init_collator(who: &T::AccountId) -> DispatchResult {
+		T::Currency::mint_into(STAKE_CURRENCY_ID, who, DEFAULT_COLLATOR_STAKE.into())?;
+		T::Rewards::deposit_stake(
+			(T::Domain::get(), STAKE_CURRENCY_ID),
+			who,
+			DEFAULT_COLLATOR_STAKE.into(),
+		)
+	}
+
+	/// Withdraw currently staked amount for target address and immediately burn it.
+	/// Disables receiving rewards onwards.
+	pub(crate) fn do_exit_collator(who: &T::AccountId) -> DispatchResult {
+		let amount = T::Rewards::account_stake((T::Domain::get(), STAKE_CURRENCY_ID), who);
+		T::Rewards::withdraw_stake((T::Domain::get(), STAKE_CURRENCY_ID), who, amount)?;
+		T::Currency::burn_from(STAKE_CURRENCY_ID, who, amount).map(|_| ())
+	}
+
+	/// Apply session changes and distribute rewards.
+	///
+	/// NOTE: Noop if any call fails.
+	fn do_advance_session() {
```

Comment:
It does not make sense to benchmark this hook as the a session change [blocks maximum weight](https://github.com/paritytech/substrate/blob/master/frame/session/src/lib.rs#L569-L572) anyways.